### DISCUSSION TITLE
AWS Deployment Stack for Langfuse v3

### DISCUSSION THREAD

**Discussion by marcklingen** [langfuse/langfuse#4645] [Ideas]
### Describe the feature or potential improvement

Currently it is recommended to use [Helm/Kubernetes](https://langfuse.com/self-hosting/kubernetes-helm) to deploy Langfuse v3. 

Let's collect interest for an AWS specific deployment stack in this thread. Please share what you'd like to see and whether you are interested in contributing to this.

### Additional information

There is [aws-samples/deploy-langfuse-on-ecs-with-fargate](https://github.com/aws-samples/deploy-langfuse-on-ecs-with-fargate) for Langfuse v2. We could aim to update it for v3.

## Edit (2025-03-31): Langfuse Terraform Module

We published https://github.com/langfuse/langfuse-terraform-aws. Please let us know your feedback on the interface and setup.

**Comment by buie** [comment]
would love a terraform module that deploys Langfuse on ECS or EKS - something similar to self hosted retool https://github.com/tryretool/retool-onpremise

**Reply by emrebsonmez** [reply]
+1 would love this

**Comment by AlinJiang** [comment]
Would love to have a CDK app to set up everything.

**Comment by LeonKalt** [comment]
Would be awsome if the aws-sample gets updated to v3

**Reply by clemra** [reply]
Thank you! We're in touch with AWS & we'll get this updated one way or another :) Will let you know here + obv. in the docs, too!

**Reply by grhaonan** [reply]
V3 is ready

**Comment by NataliaTarasovaNatoshir** [comment]
Also awaiting a CDK and AWS sample!

**Comment by moghon92** [comment]
Yes please awaiting aws sample with cdk/terraform

**Comment by tubone24** [comment]
Hello, I've created a Terraform configuration for deploying Langfuse v3 on AWS.
You can find the code at [tubone24/langfuse-v3-terraform](https://github.com/tubone24/langfuse-v3-terraform).
I hope this helps those who are looking for AWS deployment examples using Terraform.

**Reply by marcklingen** [reply]
thanks for sharing! cc @Steffen911

**Reply by preksha1511** [reply]
hey @tubone24 @marcklingen I used the above terraform code to deploy langfuse on ECS and frontend on apprunner as in the code but I am getting error "Unexpected error occurred. Please check your request and contact support: https://langfuse.com/support." and when I do a normal curl to /api/health that even gives a 404 .. I am not able to configure traces. Let me know if you need more inputs from my side

**Reply by Steffen911** [reply]
@preksha1511 Could you create an issue in the repository and link it here? 
Once you do that, please include the logs that you get on the langfuse-web component.

**Reply by preksha1511** [reply]
https://github.com/orgs/langfuse/discussions/6526 @Steffen911

**Comment by ksmin23** [comment]
Hell, I have created a CDK Python project to deploy Langfuse v3 on Amazon ECS with AWS Fargate.
Here is a github repository: https://github.com/aws-samples/deploy-langfuse-on-ecs-with-fargate/tree/main/langfuse-v3
I have written the CDK Python project with reference to the terraform code that was created by @tubone24
Thanks @tubone24 for a good example.

**Reply by ksmin23** [reply]
You can find the CDK Projects to deploy Lanfuse V2, and V3 on Amazon ECS at this: https://github.com/aws-samples/deploy-langfuse-on-ecs-with-fargate

**Reply by marcklingen** [reply]
thank you for your effort and for sharing this @ksmin23!
cc @Steffen911

**Reply by juandiegobalbi** [reply]
@ksmin23 Just what i was looking for, fantastic work with very clear instructions. Worked smoothly on the first try

**Reply by marcellewi** [reply]
@ksmin23 @marcklingen  It worked for me too! Excelent work! Thank you!

**Reply by ramammah** [reply]
Please provide an update with https endpoint for the ALB, if trace is sent over http, I dont think customers would be happy to deploy this solution

**Reply by marcklingen** [reply]
I see that you already created an issue for this here: https://github.com/aws-samples/deploy-langfuse-on-ecs-with-fargate/issues/9

**Comment by mazyu36** [comment]
Hi. I've created a project to deploy Langfuse v3 using AWS CDK with TypeScript, and I'd like to share it.
I've put some thought into it, like making it possible to change parameters per environment and more. 
Feel free to give it a try!

https://github.com/mazyu36/langfuse-with-aws-cdk

**Reply by mazyu36** [reply]
I also wrote a blog post. It covers key configuration points for each AWS resource, how to use CDK projects, and details of the troubleshooting experiences I've encountered.

https://dev.to/aws-builders/self-hosting-langfuse-v3-on-aws-using-cdk-508a

**Reply by marcklingen** [reply]
awesome, thanks for sharing!

**Reply by mazyu36** [reply]
I also added a Cognito authentication.
https://dev.to/aws-builders/adding-cognito-authentication-to-self-hosted-langfuse-with-aws-cdk-4gfe

**Comment by rkoppula** [comment]
Would love an official version from langfuse for self-hosted deployment on ECS. Makes it easy to try out in our production environment to test scalability. I have tried the local deployment for development for few days now. And I like the features and roadmap.

**Reply by Steffen911** [reply]
@rkoppula We're working on a version that runs on EKS using Fargate, but not an ECS version. To make clickhouse highly available the Kubernetes Statefulsets come in quite handy.
If you rather have ECS and do not require a highly available ClickHouse instance a setup like https://github.com/aws-samples/deploy-langfuse-on-ecs-with-fargate/tree/main/langfuse-v3 or https://github.com/mazyu36/langfuse-with-aws-cdk would be the best approach.

**Reply by nolanvenhola** [reply]
@Steffen911 EKS with Fargate isn't supported on Gov Cloud. An ECS option for Gov cloud users would be appreciated.

**Comment by Steffen911** [comment]
Hey all,
We created https://github.com/langfuse/langfuse-terraform-aws to deploy Langfuse on AWS using Terraform. We're looking for first testers to give us feedback on the interface and the setup. Please reach out here or create an issue if you have any feedback.

**Reply by dahnny012** [reply]
I wouldnt use EFS. I found abysmal performance at scale when using EFS

**Reply by Steffen911** [reply]
@dahnny012 Thank you for the input. Are there alternatives you suggest that work well with an AWS Cloud deployed Clickhouse instance?

Usually, EFS disks scale with their size or based on provisioned IO. I.e. it might be an issue with your specific configuration.

**Reply by dahnny012** [reply]
I used the defaults provided by CDK and found my queries were taking upwards of 60s for 6gbs of data. I was looking at the terraform and it doesnt seem to set the provisioned throughput nor insist on a performance strategy so likely youre using the defaults too? 

As for a solution - I have migrated to EBS https://github.com/orgs/langfuse/discussions/8675 and saw a 100x performance increase.

**Reply by dahnny012** [reply]
Its a less elegant solution as you can't just add more clickhouse nodes to scale and data is more brittle. But in the end i couldnt find a way to get good performance without paying a lot for EFS' provisioned output. (3GBs throughput is 18k a month!?)


Edit:

Just ran a test with EFS Provisioned.  I was getting 170ms (min) 500 ms (max) no mattter the provisioned throughput. This is much better than bursting mode (1000ms min 5000 ms max). I'm not sure what terraform sets the throughput mode to but you might want to explicitly set it to elastic.
